conda_libmamba_solver/shards_subset.py

```diff
@@ -234,19 +255,150 @@ def shortest_bfs(self, root_packages):
+        network_thread.start()
+
+        shardlikes_by_url = {s.url: s for s in self.shardlikes}
+        pending = set(self.nodes)  # to be sent to in_queue
+        submitted = set()  # sent to in_queue, not received from shard_out_queue
+        processed = set()  # received from shard_out_queue
+
+        def enqueue_pending():
+            log.debug(
+                "Enqueue %s pending %s",
+                len(pending),
+                sorted(node_id.package for node_id in pending),
+            )
+            shard_out_batch = []
+            shard_in_batch = []
+            # enqueue all pending nodes
+            for node_id in pending:
+                if node_id in submitted or node_id in processed:
+                    # skip already submitted or processed
+                    continue
+                # we could wind up sending duplicate node_id here?
+                shardlike = shardlikes_by_url[node_id.channel]
+                if shardlike.shard_in_memory(node_id.package):  # for monolithic repodata
+                    shard_out_batch.append((node_id, shardlike.visit_shard(node_id.package)))
+                else:
+                    shard_in_batch.append(node_id)
+                submitted.add(node_id)  # tracks everything not yet read from shard_out_batch
+            if shard_out_batch:
+                shard_out_queue.put(shard_out_batch)
+            if shard_in_batch:
+                cache_in_queue.put(shard_in_batch)
+            pending.clear()
+
+        enqueue_pending()  # initial batch
+
+        timeouts = 0
+
+        running = True
+        while running:
+            try:
+                new_shards = shard_out_queue.get(timeout=1)
+            except queue.Empty:
+                # print(f"Pending {len(pending)}, Submitted {len(submitted)}")
+                if all([x in self.nodes for x in submitted]):
+                    # print("Done but submitted not empty")
+                    running = False
+                timeouts += 1
+                # if timeouts > 10:
+                #    log.error("Timeout waiting for shard_out_queue")
+                #    raise TimeoutError("Timeout waiting for shard_out_queue")
+                continue
+            if new_shards is None:
+                running = False
+                new_shards = []
+            for node_id, shard in new_shards:
+                # add shard to appropriate ShardLike
+                submitted.remove(node_id)
+                processed.add(node_id)
+
+                parent_node = self.nodes[node_id]
+                shardlike = shardlikes_by_url[node_id.channel]
+                if node_id.package not in shardlike.visited:
+                    shardlike.visited[node_id.package] = shard
+                else:
+                    pass  # e.g. monolithic repodata is already visited
+
+                self.neighbors_2(pending, parent_node, shard)
+
+            enqueue_pending()
+            if not submitted:
```

Comment:
This end criteria is buggy, and can be seen with `python -m conda create --dry-run --override-channels -c conda-forge-sharded vaex` with an empty shards cache. Should `neighbors_2` add to submitted, or should this be "if not submitted and not pending"?